rclcpp/include/rclcpp/subscription_options.hpp

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__SUBSCRIPTION_OPTIONS_HPP_
+#define RCLCPP__SUBSCRIPTION_OPTIONS_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rclcpp/visibility_control.hpp"
+
+namespace rclcpp
+{
+
+struct SubscriptionEventCallbacks
+{
+  // NOTE(emersonknapp): empty placeholder for future additions of QoS Event callbacks
+};
+
+template<typename Alloc = std::allocator<void>>
+class SubscriptionOptions
+{
+public:
+  /// Create SubscriptionOptions with default values, optionally specifying the allocator to use.
+  /**
+   * Default values for the node options:
+   *
+   *   - deadline_callback = nullptr
+   *   - liveliness_callback = nullptr
+   *   - lifespan_callback = nullptr
+   *   - subscription_qos_profile = rmw_qos_profile_default
+   *   - ignore_local_publications = false
+   *   - allocator = nullptr
+   *
+   * \param[in] allocator allocator to use in construction of SubscriptionOptions.
+   */
+  RCLCPP_PUBLIC
+  SubscriptionOptions() = default;
+
+  RCLCPP_PUBLIC
+  explicit SubscriptionOptions(const rmw_qos_profile_t & qos_profile)
+  : qos_profile_(qos_profile) {}
+
+  /// Destructor.
+  RCLCPP_PUBLIC
+  virtual
+  ~SubscriptionOptions() = default;
+
+  /// Copy constructor.
+  RCLCPP_PUBLIC
+  SubscriptionOptions(const SubscriptionOptions & other) = default;
+
+  /// Assignment operator.
+  RCLCPP_PUBLIC
+  SubscriptionOptions &
+  operator=(const SubscriptionOptions & other) = default;
+
+
+  /// Return a reference to the subscription_qos_profile QoS.
+  RCLCPP_PUBLIC
+  const rmw_qos_profile_t &
+  qos_profile() const
+  {
+    return qos_profile_;
+  }
+
+  RCLCPP_PUBLIC
+  rmw_qos_profile_t &
+  qos_profile()
+  {
+    return qos_profile_;
+  }
+
+  /// Set the subscription_qos_profile QoS, return this for parameter idiom.
+  /**
+   * The QoS settings to be used for the subscription
+   */
+  RCLCPP_PUBLIC
+  SubscriptionOptions &
```

Comment:
Why returning this in this setter? This seems more like something we'd see in a builder class.